### PR TITLE
Solcurity Optimizations

### PR DESCRIPTION
**Pull request by Jeff-CCH**
Remove unused logic, revise license and add comments
- [x] ApproveHelper._approveZero(address,address) is never used and should be removed
- [x] Use MIT license for deployCreate3Factor.s.sol?
- [x] Use Natspec for modifiers
- [x] Explain why Signature could be reused before the deadline on the same chain or on another chain
- [x] Explain why we need to block reentrancy in comments

### CODE REVIEW

File: src/Router.sol [hunk 1]
```diff
@@ -173,6 +176,7 @@ contract Router is IRouter, EIP712, FeeGenerator {
         agent.execute{value: msg.value}(logics, tokensReturn);
     }
 
+    /// @dev Signature can be reused on any chain before the deadline
```
Comment (zodahu) [inline src/Router.sol:179]:
```suggestion
    /// @dev Allow whitelisted signers to use custom fee rules and permit the reuse of the signature until the deadline
```

- Prefer to put @dev after @notice to follow the order in https://docs.soliditylang.org/en/v0.8.20/natspec-format.html#tags
- Provide the reason of the reuse

File: src/Router.sol [hunk 2]
```diff
@@ -41,13 +41,16 @@ contract Router is IRouter, EIP712, FeeGenerator {
     /// @notice Address for invoking pause
     address public pauser;
 
+    /// @dev Modifier for blocking reentrancy by checking `currentUser`
```
Comment (zodahu) [inline src/Router.sol:44]:
```suggestion
    /// @dev Modifier for setting transient `currentUser` address and blocking reentrancy.
```

File: src/AgentImplementation.sol [hunk 3]
```diff
@@ -109,6 +109,7 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
         bool shouldChargeFeeByLogic = (callbackWithCharge & _CHARGE_MASK) != bytes32(0);
 
         // Reset immediately to prevent reentrancy
+        // If reentrancy is not blocked, anyone could utilize callback to transfer funds from agents and users
```
Comment (zodahu) [inline src/AgentImplementation.sol:112]:
```suggestion
        // If reentrancy is not blocked, an attacker could manipulate the callback contract to compel agent to execute
        // malicious logic, such as transferring funds from agents and users.
```

File: src/Router.sol [hunk 4]
```diff
@@ -41,13 +41,16 @@ contract Router is IRouter, EIP712, FeeGenerator {
...
+    ///      If reentrancy is not blocked, malicious contracts could transfer funds from agents and users
```
Comment (zodahu) [inline src/Router.sol:45]: Can you think of any attack vectors for transferring funds from the agent and user when the reentrancy check is not in place?

**Comment by ksin751119** [comment]
LGTM

File: src/Router.sol (hunk 4 above)
Comment (Jeff-CCH) [reply src/Router.sol:45]: I was thinking If the block is not placed, the flow could be router -> agent -> malicious contract -> router -> agent (transfer funds). But it only works when Router::execute get `user` from function inputs. If Router::execute takes `msg.sender` as `user`, the issue could be avoided. Suggest removing `If reentrancy ...`

File: src/Router.sol (hunk 1 above)
Comment (Jeff-CCH) [reply src/Router.sol]: fixed.

File: src/Router.sol (hunk 2 above)
Comment (Jeff-CCH) [reply src/Router.sol]: fixed.

File: src/AgentImplementation.sol (hunk 3 above)
Comment (Jeff-CCH) [reply src/AgentImplementation.sol]: fixed.

File: src/Router.sol (hunk 4 above)
Comment (zodahu) [reply src/Router.sol:45]: Agree. I think this reentrancy guard is in place to ensure that within the same transaction, the `user` does not get changed leading to the first router::execute referring to the incorrect `user`. I also haven't come up with a concrete way to exploit user's funds.

**Comment by zodahu** [comment]
LGTM

File: src/Router.sol (hunk 4 above)
Comment (Jeff-CCH) [reply src/Router.sol]: fixed.

**Pull request merged**